### DISCUSSION TITLE
Factor out the logic of moving to HEAD of upstream dependencies (gdal, proj, geos, sfcgal)

### DISCUSSION THREAD

**Discussion by rahul-ve** [postgis/docker-postgis#223] [Ideas]
Factoring out the logic of updating to the latest commit of upstream dependencies (gdal, proj, geos, sfcgal) might help minimize/avoid introducing new issues when working on a new feature or bug.

I am facing this issue right now with [PR#220](https://github.com/postgis/docker-postgis/pull/220). Was trying to add `pcre` dependency to build `address_standardizer` extension but several upstream changes also introduced new issues and made  a simple fix into a fairly complicated task for someone new.

as per @phillipross's [suggestion](https://github.com/postgis/docker-postgis/pull/220#issuecomment-766591590), starting a discussion to get feedback.
> I like this idea... or at least the goal of the idea. Without looking too closely, it seems like it would be ideal to factor out the logic for updating the heads to a separate update script. Maybe the "master" images become the most recent known to be somewhat working versions "alpha/beta" and what is known to be master now becomes "testing" or "unstable" or something like that.

**Comment by nyurik** [comment]
The [openmaptiles project](https://github.com/openmaptiles) may also need to migrate to this approach, so we are greatly interested in moving this forward. @ImreSamu suggested a similar "build with params" command in https://github.com/postgis/docker-postgis/issues/233#issuecomment-820926170: which would solve it for a generic usecase, especially because the actual docker file will still be maintained by this repo.

```bash
docker build --pull \
  \
  --build-arg BASE_PG_IMAGE=postgres:9.6 \
  \
  --build-arg SFCGAL_VERSION=v1.3.9 \
  --build-arg SFCGAL_GIT_HASH=823db7a318b8841f8296e80036ef993ddf19ebf5 \
  --build-arg PROJ_VERSION=8.0.0 \
  --build-arg PROJ_GIT_HASH=12e7a4395784d2dfb45cc0b719d37cd7e534fe91 \
  --build-arg GEOS_VERSION=3.9.1 \
  --build-arg GEOS_GIT_HASH=1b2f4bd3e19001c37846687c765595aa8c28b90f \
  --build-arg GDAL_VERSION=v3.2.2 \
  --build-arg GDAL_GIT_HASH=a33784291d19015217ea2604988e53d448e14a07 \
  --build-arg POSTGIS_VERSION=3.1.1 \
  --build-arg POSTGIS_GIT_HASH=aaf4c79a4906f11727fe6b9bccb12051a52acfe3 \
  \
  -t postgis:pg9.6-pgis3.1.1-geos3.9.1-proj8.0.0-gdal3.2.2  .
```

P.S. Related issue in OMT: https://github.com/openmaptiles/openmaptiles-tools/issues/350
